framework/src/materials/Material.C

```diff
@@ -171,8 +175,28 @@ Material::getOutputs()
 void
 Material::computeProperties()
 {
-  for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
+  // If this Material has the _constant_on_elem flag set, we take the
+  // value computed for _qp==0 and use it at all the quadrature points
+  // in the Elem.
+  if (_constant_on_elem)
+  {
+    // Compute MaterialProperty values at the first qp.
+    _qp = 0;
     computeQpProperties();
+
+    // Copy the first value computed over to the others.
+    MaterialProperties & props = _material_data->props();
+    for (auto & prop : props)
+      for (unsigned int qp = 1; qp < _qrule->n_points(); ++qp)
+        prop->qpCopy(/*to_qp=*/qp, /*rhs=*/prop, /*from_qp*/0);
```

Comment:
I am not convinced about the comments inside the function call. We do not really do this...